File: Questions/DP/minStepsToOne/minStepsToOne.py

def minSteps(n):
    dp = [-1]*(n+1)
    dp[1] = 0 
    for i in [2, n]:
        op1 = 999999        
        op2 = 999999
        op3 = dp[i-1]
        if i%3 == 0:
            op1 = dp[i//3]
        if i%2 == 0 :
            op2 = dp[i//2]
        print(min(op1, min(op2, op3)))
        dp[i] = min(op1, min(op2, op3)) + 1

    return dp[n]


n = int(input("Enter the number: "))
ans = minSteps(n)
s = "ans: " + str(ans)
print(s)

File: Questions/DP/minStepsToOne/minStepsToOne.java
import java.util.Scanner;


public class minStepsToOne {
  private static int min(int a, int b) {
    return (a>b) ? b : a;
  }
  private static int minSteps(int n) {
    int dp[] = new int[n+1] ;

    for(int i =0 ; i<n; i++) {
      dp[i] = -1 ;
    }
    dp[1] = 0 ;

    for(int i = 2; i <= n;i++) {
      int op1 = (i%2 != 0) ? Integer.MAX_VALUE : dp[i/2] ;
      int op2 = (i%3 != 0) ? Integer.MAX_VALUE : dp[i/3] ;
      int op3 = dp[i-1] ;
      dp[i] = min(op1, min(op2, op3)) + 1 ;
    }

    return dp[n] ;

  }

  public static void main(String args[]) {
    Scanner inp = new Scanner(System.in) ;
    System.out.println("Enter the number: ") ;
    int n = inp.nextInt() ;
    System.out.println("Steps: " + minSteps(n)) ;
  }
}
